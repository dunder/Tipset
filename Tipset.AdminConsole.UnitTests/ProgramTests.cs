using ApprovalTests;
using ApprovalTests.Reporters;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;
using Ninject.MockingKernel.Moq;

namespace Tipset.AdminConsole.UnitTests
{
    [TestClass]
    [UseReporter(typeof(DiffReporter))]
    public class ProgramTests
    {
        private static MoqMockingKernel mockingKernel;
        private TemporaryConsoleOutRedirection temporaryConsoleOut;
        
        [ClassInitialize]
        public static void ClassInitialize(TestContext context)
        {
            mockingKernel = new MoqMockingKernel();
            Program.Kernel = mockingKernel;
        }

        [TestInitialize]
        public void TestInitialize()
        {
            temporaryConsoleOut = new TemporaryConsoleOutRedirection(); 
        }

        [TestCleanup]
        public void TestCleanup()
        {
            temporaryConsoleOut.Dispose();
            mockingKernel.Reset();
        }

        [TestMethod]
        public void NoArgumentsDisplaysHelp()
        {
            var exitCode = Program.Main(ToCommandLineArgs(""));

            VerifyConsoleAndExitCode(exitCode, ExitCodes.InvalidArguments);
        }

        [TestMethod]
        public void HelpShouldWriteHelpToConsole()
        {
            var exitCode = Program.Main(ToCommandLineArgs("--help"));

            VerifyConsoleAndExitCode(exitCode, ExitCodes.InvalidArguments);
        }

        [TestMethod]
        public void CreateSeasonShouldCreateCommandAndExitOk()
        {
            var commandFactory = mockingKernel.GetMock<ICommandFactory>();
            var command = new Mock<ICommand>();
            commandFactory.Setup(x => x.CreateCommand(It.IsAny<string>())).Returns(command.Object);

            var exitCode = Program.Main(ToCommandLineArgs("create-season --start 2012-08-18 --end 2013-05-18"));

            Assert.AreEqual((int)ExitCodes.Ok, exitCode);
            commandFactory.Verify(cf => cf.CreateCommand("create-season"));
        }

        [TestMethod]
        public void CreatePlayerShouldCreateCommandAndExitOk()
        {
            var commandFactory = mockingKernel.GetMock<ICommandFactory>();
            var command = new Mock<ICommand>();
            commandFactory.Setup(x => x.CreateCommand(It.IsAny<string>())).Returns(command.Object);

            var exitCode = Program.Main(ToCommandLineArgs("create-player --name Matias --email matiasjansson@gmail.com"));

            Assert.AreEqual((int)ExitCodes.Ok, exitCode, "unexpected exit code");
            commandFactory.Verify(cf => cf.CreateCommand("create-player"));
        }

        private void VerifyConsoleAndExitCode(int exitCode, ExitCodes expectedExitCode = ExitCodes.Ok)
        {
            Assert.AreEqual((int) expectedExitCode, exitCode, "unexpected exit code");
            Approvals.Verify(temporaryConsoleOut.TemporaryOut);
        }

        private static string[] ToCommandLineArgs(string commandLine)
        {
            return commandLine.Split(' ');
        }
    }
}
