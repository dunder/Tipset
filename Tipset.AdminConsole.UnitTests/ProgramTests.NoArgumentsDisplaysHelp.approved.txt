CommandLine 1.9.71-stable
Copyright (c) 2005 - 2013 Giacomo Stelluti Scala

  --create-season    yyyy-MM-dd

  --create-player    

  --help             

  --help             


